executor/foreign_key.go

```diff
@@ -0,0 +1,495 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"sync/atomic"
+
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/parser/model"
+	plannercore "github.com/pingcap/tidb/planner/core"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/set"
+)
+
+// WithForeignKeyTrigger indicates the executor has foreign key check or cascade.
+type WithForeignKeyTrigger interface {
+	GetFKChecks() []*FKCheckExec
```

Comment:
Only `InsertExec` implements this interface, so other statements like `update` won't check foreign keys, is it ok?

Reply:
That looks to be the case. Here is the testcase from the [MySQL manual](https://dev.mysql.com/doc/mysql-tutorial-excerpt/8.0/en/example-foreign-keys.html):

```sql
DROP TABLE IF EXISTS parent, child;
SET foreign_key_checks = 1;
SET GLOBAL tidb_enable_foreign_key=1;
CREATE TABLE parent (
    id INT NOT NULL,
    PRIMARY KEY (id)
) ENGINE=INNODB;
CREATE TABLE child (
    id INT,
    parent_id INT,
    INDEX par_ind (parent_id),
    FOREIGN KEY (parent_id)
        REFERENCES parent(id)
) ENGINE=INNODB;
INSERT INTO parent (id) VALUES (1);
INSERT INTO child (id,parent_id) VALUES (1,1);
INSERT INTO child (id,parent_id) VALUES(2,2); # error expected (works)
DELETE FROM parent WHERE id = 1; # error expected (not supported yet)
```

May I suggest updating the PR title to just be for `INSERT` and not DML?

Reply:
Great, currently this PR only supports foreign key check in `INSERT`.
Before the commit [403aaa89fde08f2238a80dca4eb44bd6e98b916b](https://github.com/pingcap/tidb/pull/37466/commits/403aaa89fde08f2238a80dca4eb44bd6e98b916b), this PR was support foreign key check in `INSERT/UPDATE/DELETE`, but the PR is so large, then I only support foreign key check in `INSERT`.